hail/python/hail/backend/service_backend.py

```diff
@@ -574,32 +573,8 @@ async def inputs(infile, _):
             await write_str(infile, self.billing_project)
             await write_str(infile, self.remote_tmpdir)
             await write_str(infile, name)
-
-        if name in BUILTIN_REFERENCE_DOWNLOAD_LOCKS:
-            with BUILTIN_REFERENCE_DOWNLOAD_LOCKS[name]:
-                try:
-                    with open(Path(self.user_local_reference_cache_dir, name)) as f:
-                        return orjson.loads(f.read())
-                except FileNotFoundError:
-                    _, resp, _ = await self._rpc('get_reference(...)', inputs, progress=progress)
-                    with open(Path(self.user_local_reference_cache_dir, name), 'wb') as f:
-                        f.write(resp)
-        else:
-            _, resp, _ = await self._rpc('get_reference(...)', inputs)
-
-        return orjson.loads(resp)
-
-    def get_references(self, names):
-        return async_to_blocking(self._async_get_references(names))
-
-    async def _async_get_references(self, names, *, progress: Optional[BatchProgressBar] = None):
-        if progress is None:
-            with BatchProgressBar() as progress:
-                return await asyncio.gather(*[self._async_get_reference(name, progress=progress)
-                                              for name in names])
-        else:
-            return await asyncio.gather(*[self._async_get_reference(name, progress=progress)
-                                          for name in names])
+        _, reference, _ = await self._rpc('get_reference(...)', inputs, progress=progress)
+        return orjson.loads(reference)
```

Comment:
This now has no effect. The ServiceBackend currently lacks support for custom references, so the get_reference RPC can only return one of the default references. I think we should eliminate all this and just raise an error that no such reference exists.

This PR is a clear improvement that I want to land posthaste.

However, we should also pull this thread fully taught: references should be owned by Python. The Scala `ReferenceGenome.references` map should just be a cache for the Java Backends (Spark and Local). The Service Backend can't use `ReferenceGenome.references` for user references anyway because the `ReferenceGenome` class is in the JAR and thus shared across users.

References should live in a map on the Python `Backend` class. The Py4JBackend should also update the Java-side when a reference is added. The ServiceBackend should send along every custom reference genome with every RPC (they're not very large, but, yes, we should find a way to avoid sending them when not necessary; not necessary for MVP).

Reply:
cc: @tpoterba I want to move references into Python, see plan above, curious if you have any more to add.